### DISCUSSION TITLE
Subscription with non-working callback

### DISCUSSION THREAD

**Discussion by Xindictus** [gstavrinos/ai_msc_robotic_systems_assignments#43] [Q&A]
Με αφορμή το σχόλιο εδώ https://github.com/gstavrinos/ai_msc_robotic_systems_assignments/discussions/42#discussioncomment-9523964, ανοίγω νέο thread.

Έχω prints πριν και μετά το subscription, όπως και στην callback. Το μόνο print που δε γίνεται είναι αυτό μέσα στην callback.

To `ros2 topic echo` μου δίνει κανονικά msg.

**Comment by gstavrinos** [comment]
Λογικά φτιάχνεις κάπου τον subscriber και γίνεται desctruct αυτόματα καθώς βγαίνει out of scope. Δοκίμασε να τον κάνεις class member. Αν θέλεις στείλει και το κομμάτι κώδικα που τον αρχικοποιείς.

**Reply by Xindictus** [reply]
```python
def on_activate(self, state: State) -> TransitionCallbackReturn:
    self.get_logger().info("Activating...")

    self.__subscribe_to_amcl()

    return super().on_activate(state)

def __subscribe_to_amcl(self) -> None:
    self.__log('Subscribing to AMCL...')

    # Create a QoS profile matching the publisher's QoS settings
    qos_profile = QoSProfile(
        reliability=QoSReliabilityPolicy.RELIABLE,
        history=QoSHistoryPolicy.KEEP_LAST,
        depth=100,
        durability=QoSDurabilityPolicy.VOLATILE,
        lifespan=rclpy.duration.Duration(seconds=0),
        deadline=rclpy.duration.Duration(seconds=0),
        liveliness=QoSLivelinessPolicy.AUTOMATIC,
        liveliness_lease_duration=rclpy.duration.Duration(seconds=0)
    )

    self.subscription = self.create_subscription(
        PoseWithCovarianceStamped,
        '/amcl_pose',
        self.__amcl_callback,
        qos_profile
    )

    self.__log('Subscribed to AMCL...')

def __amcl_callback(self, msg: PoseWithCovarianceStamped) -> None:
    self.get_logger().info(f'Initial pose set: {msg}')
```

**Reply by gstavrinos** [reply]
Φαντάζομαι ότι και χωρίς το qos πάλι το ίδιο κάνει. Ωστόσο, δεν βλέπω κάτι περίεργο... Δοκίμασε να το τρέξεις και ξεχωριστά σε ένα άλλο terminal να δεις αν σου βγάζει κάτι που χάνεις μέσα στα outputs από όλα τα nodes μαζεμένα.

Όπως και να 'χει, δοκίμασε και τη λύση που σου είπα με τα transform listeners, μήπως σε ξεκολλήσει.

**Reply by Xindictus** [reply]
> Φαντάζομαι ότι και χωρίς το qos πάλι το ίδιο κάνει.

Yeap. Επίσης δοκιμάσα και με values 10 & 100, μιας και με `ros2 topic info --verbose`, ο `publisher` δούλευε με 100 αλλά και πάλι δεν έπαιξε. Είναι σαν να μην έπαιρνε μηνύματα, γιατί κατά τ'άλλα το execution προχωρούσε.

> Δοκίμασε να το τρέξεις και ξεχωριστά σε ένα άλλο terminal να δεις αν σου βγάζει κάτι που χάνεις μέσα στα outputs από όλα τα nodes μαζεμένα.

Done already. Τα ίδια. Επίσης δοκίμασα το topic name με και χωρίς slash στην αρχή.
